lib/logstash/outputs/elasticsearch.rb

```diff
@@ -472,4 +504,42 @@ def teardown
     require "logstash/outputs/elasticsearch/#{name}"
   end
 
+  def retry_flush(teardown=false)
+    @retry_flush_mutex.lock
+    begin
+      @retry_queue_needs_flushing.wait(@retry_flush_mutex) unless teardown
+      unless @retry_queue.empty?
+        buffer = @retry_queue.size.times.map do
+          next_action, next_doc, next_event = @retry_queue.pop
+          next_event['@metadata']['retry_count'] += 1
+
+          if next_event['@metadata']['retry_count'] > @max_retries
+            @logger.error "too many attempts at sending event. dropping: #{next_event}"
+            return nil
+          end
+
+          [next_action, next_doc, next_event]
+        end.compact
+        submit(buffer)
+      end
+    ensure
+      @retry_flush_mutex.unlock
+      @retry_queue_not_full.signal if @retry_queue.size < @retry_max_items
+    end
+  end
+
+  
+  def retry_push(action)
+    @retry_queue << action
+    @retry_queue_needs_flushing.signal if @retry_queue.size >= @retry_max_items
+  end
```

Comment:
I suggest `retry_push` should also support a list of actions parameter to avoid calling `.signal` on all actions once `@retry_queue.size >= @retry_max_items`. 

Calling `retry_flush` could now be
```ruby
      unless failed_actions.empty?
        @logger.debug "#{failed_actions.size}/#{actions.size} events were unsuccessful in sending"
        retry_push(failed_actions)
      end
```

and `retry_flush` could handle single or multiple actions with something like:

```ruby
  def retry_push(actions)
    Array(actions).each{|action| @retry_queue << action}
    @retry_queue_needs_flushing.signal if @retry_queue.size >= @retry_max_items
  end
```